src/app/clusters/account-login-server/account-login-server.cpp

```diff
@@ -39,18 +39,51 @@
  ******************************************************************************/
 
 #include <app/Command.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/common/gen/command-id.h>
 #include <app/util/af.h>
+#include <string>
 
-bool emberAfAccountLoginClusterGetSetupPINCallback(chip::app::Command * commandObj, unsigned char *)
+bool accountLoginClusterIsUserLoggedIn(std::string requestTempAccountIdentifier, std::string requestSetupPin);
+std::string accountLoginClusterGetSetupPin(std::string requestTempAccountIdentifier, chip::EndpointId endpoint);
+
+void sendResponse(chip::app::Command * command, const char * responseSetupPin)
 {
-    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
-    emberAfSendImmediateDefaultResponse(status);
+    CHIP_ERROR err                         = CHIP_NO_ERROR;
+    chip::app::CommandPathParams cmdParams = { emberAfCurrentEndpoint(), /* group id */ 0, ZCL_ACCOUNT_LOGIN_CLUSTER_ID,
+                                               ZCL_GET_SETUP_PIN_RESPONSE_COMMAND_ID,
+                                               (chip::app::CommandPathFlags::kEndpointIdValid) };
+    chip::TLV::TLVWriter * writer          = nullptr;
+    SuccessOrExit(err = command->PrepareCommand(cmdParams));
+    VerifyOrExit((writer = command->GetCommandDataElementTLVWriter()) != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    SuccessOrExit(err = writer->PutString(chip::TLV::ContextTag(0), responseSetupPin));
+    SuccessOrExit(err = command->FinishCommand());
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Failed to encode GetSetupPIN command. Error:%s", chip::ErrorStr(err));
+    }
+}
+
+bool emberAfAccountLoginClusterGetSetupPINCallback(chip::app::Command * command, unsigned char * tempAccountIdentifier)
+{
+    std::string tempAccountIdentifierString(reinterpret_cast<char *>(tempAccountIdentifier));
+    std::string responseSetupPin = accountLoginClusterGetSetupPin(tempAccountIdentifierString, emberAfCurrentEndpoint());
+    sendResponse(command, responseSetupPin.c_str());
     return true;
 }
 
-bool emberAfAccountLoginClusterLoginCallback(chip::app::Command * commandObj, unsigned char *, unsigned char *)
+bool emberAfAccountLoginClusterLoginCallback(chip::app::Command * command, unsigned char * tempAccountIdentifier,
+                                             unsigned char * tempSetupPin)
 {
-    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    std::string tempAccountIdentifierString(reinterpret_cast<char *>(tempAccountIdentifier));
+    std::string tempSetupPinString(reinterpret_cast<char *>(tempSetupPin));
+    bool isLoggedIn      = accountLoginClusterIsUserLoggedIn(tempAccountIdentifierString, tempSetupPinString);
+    EmberAfStatus status = isLoggedIn ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_NOT_AUTHORIZED;
+    if (!isLoggedIn)
+    {
+        ChipLogError(Zcl, "User is not authorized. Error:%s", chip::ErrorStr(EMBER_ZCL_STATUS_NOT_AUTHORIZED));
```

Comment:
This doesn't really make sense.  `ErrorStr` takes a CHIP_ERROR, not an EMBER_* status....  What it will do with this value is pretty random and can't be relied on.

Reply:
Removed the `Error: %s...`